dynamic-libraries.md

```diff
@@ -0,0 +1,312 @@
+  (:literal "/* functions */")
+  (:function
+   (call-callback :void ((fn :callback) (out_buffer :char-buffer)))))
+
+(define-aggregate-library libcallback (:function-linkage "CALLBACKING_API")
+  sbcl-librarian:handles sbcl-librarian:environment libcallback-api)
+~~~
+
+`define-api` outlines the structure of the library code to be created, specifying the error map, types, functions, and their order (`:literal` is used for comments in this case). The function `call-callback` uses previously defined types for its arguments: the `callback` type for the first argument named `fn` and the `:char-buffer` type for its second argument `out_buffer`. The `:function-prefix` option means the actual name of the exported function will be `callback_call_callback`.
+
+`define-aggregate-library` defines the entire library, specifying what should be included and in what order.
+
+### Compile LISP Code
+
+Now you can compile the Lisp code and generate the C sources for compiling the library and the Python wrapper.
+
+Set a couple of environment variables for convenience:
+
+~~~bash
+# Directory with SBCL sources
+export SBCL_SRC=~/.roswell/src/sbcl-2.4.1
+# Directory with this project, don't forget the double slash at the end
+# or it might not work
+export CL_SOURCE_REGISTRY="~/prg/sbcl-librarian//"
+~~~
+
+Libraries are usually not searched for in the current directory on more modern Linux-based systems, similar to paths Python searches for libraries.
+
+~~~bash
+export LD_LIBRARY_PATH=.:
+export PATH=.:$PATH
+~~~
+
+`script.lisp` is a straightforward Lisp script for compiling the Lisp sources and outputting the wrapper code and the Lisp core.
+
+~~~lisp
+(require '#:asdf)
+
+(asdf:load-system '#:libcallback)
+
+(in-package #:sbcl-librarian/example/libcallback)
+
+(build-bindings libcallback ".")
+(build-python-bindings libcallback ".")
+(build-core-and-die libcallback "." :compression t)
+~~~
+
+Now you have a couple of new files.
+
+`libcallback.c` is the source code for the library:
+
+~~~c
+#define CALLBACKING_API_BUILD
+
+#include "libcallback.h"
+
+void (*lisp_release_handle)(void* handle);
+int (*lisp_handle_eq)(void* a, void* b);
+void (*lisp_enable_debugger)();
+void (*lisp_disable_debugger)();
+void (*lisp_gc)();
+err_t (*callback_call_callback)(void* fn, char* out_buffer);
+
+extern int initialize_lisp(int argc, char **argv);
+
+CALLBACKING_API int init(char* core) {
+  static int initialized = 0;
+  char *init_args[] = {"", "--core", core, "--noinform", };
+  if (initialized) return 1;
+  if (initialize_lisp(4, init_args) != 0) return -1;
+  initialized = 1;
+  return 0; }
+~~~
+
+At the top, you'll find several SBCL-related functions, such as `lisp_gc`, which signals to the Lisp garbage collector that it is a good time to run. Then there is a pointer to the `callback_call_callback` function. Finally, the `init` function, which should be run before executing any Lisp code.
+
+Currently, there is no way to de-initialize the Lisp core.
```

Comment:
Add a timestamp or mention somewhere (github issue) to track a potential update on this? (if it's a goal and being worked on)

Reply:
Hi, I'm not sure whether they plan to add this functionality. I'm not even sure whether to include this sentence or not, honestly ...